rfcs/9999-new-merkle-storage.md

```diff
@@ -0,0 +1,140 @@
+the data with the associated hash. The `Ref` type has two parameters, A) a
+phantom that is either `full` or `partial` indicating whether we are guaranteed
+to have all the referenced data as well, and B) what the type of the data is.
+There is a function that takes a `Ref` and ensures that it and all its
+references are stored on disk. We call this "pinning" the `Ref`.
+
+In the table below, the rows correspond to whether the data is pinned and the
+columns correspond to the value of the "fullness" type parameter. The inner
+elements are what those combinations are in the current system (for ledgers).
+
+|        | Full                             | Partial         |
+|--------|----------------------------------|-----------------|
+| Disk   | `Merkle_ledger.Database`         | N/A             |
+| Memory | `Ledger.Mask.Attached` (to null) | `Sparse_ledger` |
+
+Most (maybe all?) of the in-memory ledgers we use in the daemon are actually
+masks on top of a database ledger, which doesn't fit into the new model
+described in this RFC.
+
+The system for synchronizing Merkle structures over the network exists at this
+level, replacing `syncable_ledger`.
+
+On top of the underlying storage layer are the actual Merkle structures. In
+order to store an Merkle object, the programmer writes a module that describes
+the object, how to (de)serialize it, and how to find its references. They pass
+that object into a functor and get the `Ref` machinery, specialized to the type
+in question, back. When implementing this, we'd write a module that describes
+the complete binary tree that is our ledger - the result of the functor would be
+used to build something close to `Base_ledger_intf.S`. We could also use this
+for a Merkleized parallel scan state, or to replace
+`receipt_chain_database`.
+
+### Storage management
+
+There are two tables mapping hashes to objects: an in-memory hashtable and
+RocksDB on disk. The in-memory table contains *weak* references, and entries are
+removed from it by a finalizer when the associated data is garbage collected. We
+can rely on OCaml's garbage collector for the in-memory data, but for the on
+disk data we'll need to use reference counting and delete objects from the
+database when their number of incoming pinned references goes to zero. (If they
+have >0 memory references and 0 pinned references they need to be read from disk
+into memory before being deleted from disk.)
+
+A `Ref` can be constructed by providing the data it corresponds to and either
+the hashes of or `Ref`s to anything it references. Or, you can look them up by
+hash, and get a `Ref` back if it's already stored. When a `Ref` is created anew,
+the object is *not* hashed immediately. It is instead added to a queue of
+unhashed data that is discharged every major GC by hashing the items that are
+still live and adding them to the hashtable. The principle here is that we may
+make many changes to a structure in sequence and only care about the hash at the
```

Comment:
It's a nice principle. Do we have this usage pattern in our daemon, though? Each transaction job in the scan state needs the Merkle paths of the updated accounts, which means we can't avoid any of the intermediate hashes I think this would save us...

Reply:
Yeah, I agree that I don't think this will help us avoid any hashes. I do think it could allow us to share the computation of hashes better than we do right now, at least.

Reply:
We do it when we apply transactions in a staged ledger. It's most of why the new staged ledger tests are so slow. You're right that the scan state needs the updates, so I suppose staged ledger is the only current use case.